src/lib/get_cmdline.c

```diff
@@ -306,3 +306,406 @@ int dump_fd_info(const char *dest_filename, const char *proc_pid_fd_path)
+        while ((c = fgetc(fin)) != EOF && (pre_c == '\\' || c != ' ') && c != '\n')
+            ;
+
+        if (c == EOF && fn != (ARRAY_SIZE(mntnf->mntnf_items) - 1))
+        {
+            log_notice("Unexpected end of file");
+            r = 1;
+            goto get_mount_info_cleanup;
+        }
+
+        /* we are standing on ' ', so len is +1 longer than the string we want to copy*/
+        const long pos_cur = ftell(fin);
+        if (pos_cur < 0)
+        {
+            pwarn_msg("ftell");
+            r = -1;
+            goto get_mount_info_cleanup;
+        }
+
+        size_t len = (pos_cur - pos_bck);
+        mntnf->mntnf_items[fn] = xmalloc(sizeof(char) * (len));
+
+        --len; /* we are standing on ' ' */
+
+        if (fseek(fin, pos_bck, SEEK_SET) < 0)
+        {
+            pwarn_msg("fseek");
+            goto get_mount_info_cleanup;
+        }
+
+        if (fread(mntnf->mntnf_items[fn], sizeof(char), len, fin) != len)
+        {
+            pwarn_msg("fread");
+            goto get_mount_info_cleanup;
+        }
+
+        mntnf->mntnf_items[fn][len] = '\0';
+
+        if (fseek(fin, 1, SEEK_CUR) < 0)
+        {
+            pwarn_msg("fseek");
+            goto get_mount_info_cleanup;
+        }
+
+        /* ignore optional fields
+           'shared:X' 'master:X' 'propagate_from:X' 'unbindable'
+         */
+        if (   strncmp("shared:", mntnf->mntnf_items[fn], strlen("shared:")) == 0
+            || strncmp("master:", mntnf->mntnf_items[fn], strlen("master:")) == 0
+            || strncmp("propagate_from:", mntnf->mntnf_items[fn], strlen("propagate_from:")) == 0
+            || strncmp("unbindable", mntnf->mntnf_items[fn], strlen("unbindable")) == 0)
+        {
+            free(mntnf->mntnf_items[fn]);
+            mntnf->mntnf_items[fn] = NULL;
+            continue;
+        }
+
+        ++fn;
+    }
+
+get_mount_info_cleanup:
+    if (r)
+        mountinfo_destroy(mntnf);
+
+    return r;
+}
+
+static int proc_ns_eq(struct ns_ids *lhs_ids, struct ns_ids *rhs_ids, int neg)
+{
+    for (size_t i = 0; i < ARRAY_SIZE(lhs_ids->nsi_ids); ++i)
+        if (    lhs_ids->nsi_ids[i] != PROC_NS_UNSUPPORTED
+             && (neg ? lhs_ids->nsi_ids[i] == rhs_ids->nsi_ids[i]
+                     : lhs_ids->nsi_ids[i] != rhs_ids->nsi_ids[i]))
+            return 1;
+
+    return 0;
+}
+
+static int get_process_ppid(pid_t pid, pid_t *ppid)
+{
+    int r = 0;
+    static char stat_path[sizeof("/proc/%lu/stat") + sizeof(long)*3];
+    sprintf(stat_path, "/proc/%lu/stat", (long)pid);
+
+    FILE *stat_file = fopen(stat_path, "re");
+    if (stat_file == NULL)
+    {
+        pwarn_msg("Failed to open stat file");
+        r = -1;
+        goto get_process_ppid_cleanup;
+    }
+
+    int p = fscanf(stat_file, "%*d %*s %*c %d", ppid);
+    if (p != 1)
+    {
+        log_notice("Failed to parse stat line %d\n", p);
+        r = -2;
+        goto get_process_ppid_cleanup;
+    }
+
+get_process_ppid_cleanup:
```

Comment:
Nitpick: labels are local to functions so you don't have to have unique names (not a problem if you have, though).